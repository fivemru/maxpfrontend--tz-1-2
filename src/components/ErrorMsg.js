import React from 'react';
import PropTypes from 'prop-types';

export const ErrorMsg = props => {
  const { error } = props;

  return <div className="error">{error}</div>;
};

ErrorMsg.propTypes = {
  error: PropTypes.string.isRequired,
};
