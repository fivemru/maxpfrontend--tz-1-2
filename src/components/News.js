import React from 'react';
import './News.css';

export const News = props => {
  const { className, title, image, date } = props;

  const [dateFormatted] = date.split('T');

  return (
    <div className={`news ${className}`}>
      <img
        className='news__image'
        src={image}
        width='128'
        height='128'
        alt={title}
      />
      <div className='news__date'>{dateFormatted}</div>
      <h2 className='news__title'>{title}</h2>
    </div>
  );
};
