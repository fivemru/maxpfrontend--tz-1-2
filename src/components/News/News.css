.news {
  padding: 20px;
  border: 1px solid #ccc;
  border-radius: 3px;
  background-color: #fff;
}

.news__title {
  margin: 0;
  padding-bottom: 10px;
  font-size: 18px;
}
