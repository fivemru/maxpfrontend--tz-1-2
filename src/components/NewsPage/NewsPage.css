.news-list {
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  flex-wrap: wrap;
}

.news-list__item {
  margin: 10px 0;
}

.news-list__item:last-child {
  margin-bottom: 0;
}
