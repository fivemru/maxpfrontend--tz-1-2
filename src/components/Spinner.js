import React from 'react';

export const Spinner = props => {
  const { text = 'Loading...' } = props;

  return <div>{text}</div>;
};
