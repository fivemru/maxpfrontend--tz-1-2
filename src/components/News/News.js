import React from 'react';
import PropTypes from 'prop-types';
import './News.css';

export const News = props => {
  const { className, id, title, text } = props;

  return (
    <div className={`news ${className}`}>
      <h2 className='news__title'>
        #{id} {title}
      </h2>
      <div className='news__text'>{text}</div>
    </div>
  );
};

News.propTypes = {
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired
};
