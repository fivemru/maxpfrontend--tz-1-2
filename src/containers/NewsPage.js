import { connect } from 'react-redux';
import { NewsPage } from '../components/NewsPage';
import { getNews } from '../actions/news';

const mapStateToProps = ({ news }) => ({
  isPending: news.isPending,
  error: news.error,
  data: news.data
});

export default connect(
  mapStateToProps,
  { getNews }
)(NewsPage);
