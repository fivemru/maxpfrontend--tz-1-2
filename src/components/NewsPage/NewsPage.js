import React, { useEffect } from 'react';
// import React from 'react';
import PropTypes from 'prop-types';
import { News } from '../News';
import { Spinner } from '../Spinner';
import { ErrorMsg } from '../ErrorMsg';
import './NewsPage.css';

export const NewsPage = props => {
  const { isPending, data, getNews, error } = props;

  useEffect(() => {
    getNews();
  }, []);

  return (
    <div>
      <h1>News page</h1>
      {error && <ErrorMsg error={error} />}
      {isPending && <Spinner text="Loading news..." />}
      {data && (
        <>
          <div>
            Total news: <strong>{data.length}</strong>
          </div>

          <div className="news-list">
            {data.map(news => (
              <News
                key={news.id}
                id={news.id}
                title={news.title}
                text={news.text}
                className="news-list__item"
              />
            ))}
          </div>
        </>
      )}
    </div>
  );
};

NewsPage.propTypes = {
  isPending: PropTypes.bool.isRequired,
  error: PropTypes.string,
  getNews: PropTypes.func.isRequired,
  data: PropTypes.array,
};
