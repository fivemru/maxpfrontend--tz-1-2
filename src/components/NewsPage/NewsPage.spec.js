import React from 'react';
import { shallow } from 'enzyme';
import { render } from 'react-testing-library';
import { NewsPage } from './NewsPage';
import { ErrorMsg } from '../ErrorMsg';
import { Spinner } from '../Spinner';
import { News } from '../News';

describe('News page', () => {
  const initProps = {
    isPending: false,
    error: null,
    data: [],
    getNews: () => {},
  };

  describe('initial', () => {
    const mockGetNews = jest.fn();
    const props = { ...initProps, getNews: mockGetNews };

    const newsPage = shallow(<NewsPage {...props} />);

    it('not render <ErrorMsg />', () => {
      expect(newsPage.find(ErrorMsg)).toHaveLength(0);
    });

    it('not render <Spinner />', () => {
      expect(newsPage.find(Spinner)).toHaveLength(0);
    });

    it('not render News', () => {
      expect(newsPage.find(News)).toHaveLength(0);
    });

    it('render <NewsPage />', () => {
      expect(newsPage).toHaveLength(1);
    });

    it('getNews has been called 1 times', () => {
      render(<NewsPage {...props} />);
      expect(mockGetNews).toBeCalledTimes(1);
    });
  });

  describe('pending', () => {
    const props = { ...initProps, isPending: true };

    const newsPage = shallow(<NewsPage {...props} />);

    it('not render <ErrorMsg />', () => {
      expect(newsPage.find(ErrorMsg)).toHaveLength(0);
    });

    it('render <Spinner />', () => {
      expect(newsPage.find(Spinner)).toHaveLength(1);
    });

    it('not render <News />', () => {
      expect(newsPage.find(News)).toHaveLength(0);
    });
  });

  describe('has error', () => {
    const props = { ...initProps, error: 'some error' };

    const newsPage = shallow(<NewsPage {...props} />);

    it('render <ErrorMsg />', () => {
      // expect(newsPage.find(ErrorMsg).prop('error')).toEqual(props.error);
      expect(
        newsPage
          .find(ErrorMsg)
          .shallow()
          .text()
      ).toEqual(props.error);
    });

    it('not render <Spinner />', () => {
      expect(newsPage.find(Spinner)).toHaveLength(0);
    });

    it('not render <News />', () => {
      expect(newsPage.find(News)).toHaveLength(0);
    });

    it('toMatchSnapshot', () => {
      expect(newsPage).toMatchSnapshot();
    });
  });

  describe('has news', () => {
    const props = {
      ...initProps,
      data: [
        { id: 1, title: 'some title 1', text: 'some text 1' },
        { id: 2, title: 'some title 2', text: 'some text 2' },
      ],
    };

    const newsPage = shallow(<NewsPage {...props} />);

    it('not render <ErrorMsg />', () => {
      expect(newsPage.find(ErrorMsg)).toHaveLength(0);
    });

    it('not render <Spinner />', () => {
      expect(newsPage.find(Spinner)).toHaveLength(0);
    });

    it('rendered 2 <News />', () => {
      expect(newsPage.find(News)).toHaveLength(2);
    });
  });
});
