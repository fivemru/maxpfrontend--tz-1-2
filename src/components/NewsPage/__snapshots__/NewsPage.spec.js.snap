// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News page has error toMatchSnapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewsPage
    data={Array []}
    error="some error"
    getNews={[Function]}
    isPending={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          News page
        </h1>,
        <ErrorMsg
          error="some error"
        />,
        false,
        <React.Fragment>
          <div>
            Total news: 
            <strong>
              0
            </strong>
          </div>
          <div
            className="news-list"
          />
        </React.Fragment>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "News page",
        },
        "ref": null,
        "rendered": "News page",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "error": "some error",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <div>
              Total news: 
              <strong>
                0
              </strong>
            </div>,
            <div
              className="news-list"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Total news: ",
                <strong>
                  0
                </strong>,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Total news: ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": 0,
                },
                "ref": null,
                "rendered": 0,
                "type": "strong",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "news-list",
            },
            "ref": null,
            "rendered": Array [],
            "type": "div",
          },
        ],
        "type": Symbol(react.fragment),
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            News page
          </h1>,
          <ErrorMsg
            error="some error"
          />,
          false,
          <React.Fragment>
            <div>
              Total news: 
              <strong>
                0
              </strong>
            </div>
            <div
              className="news-list"
            />
          </React.Fragment>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "News page",
          },
          "ref": null,
          "rendered": "News page",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "error": "some error",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <div>
                Total news: 
                <strong>
                  0
                </strong>
              </div>,
              <div
                className="news-list"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Total news: ",
                  <strong>
                    0
                  </strong>,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Total news: ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 0,
                  },
                  "ref": null,
                  "rendered": 0,
                  "type": "strong",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [],
                "className": "news-list",
              },
              "ref": null,
              "rendered": Array [],
              "type": "div",
            },
          ],
          "type": Symbol(react.fragment),
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
